src/mempool/storage.rs

```diff
@@ -1,45 +1,161 @@
 use anyhow::{bail, Result};
 use bincode::{Decode, Encode};
+use indexmap::IndexMap;
 use serde::{Deserialize, Serialize};
 use sha3::{Digest, Sha3_256};
-use std::{
-    collections::{BTreeSet, HashMap},
-    fmt::Display,
-    io::Write,
-    vec,
-};
+use std::{collections::HashMap, fmt::Display, hash::Hash, vec};
 use tracing::{debug, error, warn};
 
 use crate::{
     model::{Hashable, Transaction, TransactionData, ValidatorPublicKey},
     node_state::NodeState,
+    p2p::network::SignedByValidator,
 };
 
+use super::MempoolNetMessage;
+
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub enum DataProposalVerdict {
     Empty,
-    Wait(Option<CarHash>),
+    Wait(Option<DataProposalHash>),
     Vote,
-    DidVote,
     Refuse,
 }
 
-pub type Cut = Vec<(ValidatorPublicKey, CarHash)>;
+#[derive(Clone, Debug, Default, Serialize, Deserialize, Encode, Decode, Eq, PartialEq)]
+pub struct DataProposal {
+    pub id: u32,
+    pub parent_data_proposal_hash: Option<DataProposalHash>,
+    pub txs: Vec<Transaction>,
+}
+
+#[derive(Debug, Clone, Encode, Decode, PartialEq, Eq, Serialize, Deserialize)]
+pub struct LaneEntry {
+    pub data_proposal: DataProposal,
+    pub signatures: Vec<SignedByValidator<MempoolNetMessage>>,
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct Lane {
+    pub last_cut: Option<DataProposalHash>,
+    pub data_proposals: IndexMap<DataProposalHash, LaneEntry>,
+    pub waiting: Vec<DataProposal>,
+}
+
+impl Encode for Lane {
```

Comment:
Peut-être que pour éviter d'implémenter Encode/Decode pour chaque struct qui l'utilise, tu pourrais les implémenter de manière générique:
`impl <Key, Value> Encode for IndexMap<Key, Value> {...}` et pareil pour Decode

Reply:
Y'aura pas de problème avec l'orphan rule ? :thinking: